crates/runtime/benches/bench.rs

```diff
@@ -0,0 +1,127 @@
+//! This is a benchmark test suite for the Spice runtime.
+//!
+//! It performs the following actions:
+//! 1. Starts the runtime with all of the datasets to test loaded.
+//! 2. Runs a series of queries against the runtime.
+//! 3. Reports the results to the spice.ai dataset <https://spice.ai/spicehq/spice-tests/datasets/oss_benchmarks>
+
+// spice.ai/spicehq/spice-tests/datasets/spicehq."spice-tests".oss_benchmarks
+// schema
+// run_id, started_at, finished_at, query_name, status, min_duration, max_duration, iterations, commit_sha
+
+use std::sync::Arc;
+
+use arrow::array::RecordBatch;
+use datafusion::datasource::provider_as_source;
+use datafusion::logical_expr::{LogicalPlanBuilder, UNNAMED_TABLE};
+use datafusion::{dataframe::DataFrame, datasource::MemTable, execution::context::SessionContext};
+use results::BenchmarkResultsBuilder;
+use runtime::{dataupdate::DataUpdate, Runtime};
+
+use crate::results::Status;
+
+mod results;
+mod setup;
+
+mod bench_spicecloud;
+
+#[tokio::main]
+async fn main() -> Result<(), String> {
+    let mut upload_results_dataset: Option<String> = None;
+    if let Ok(env_var) = std::env::var("UPLOAD_RESULTS_DATASET") {
+        println!("UPLOAD_RESULTS_DATASET: {env_var}");
+        upload_results_dataset = Some(env_var);
+    }
+
+    let (mut benchmark_results, mut rt) = setup::setup_benchmark(&upload_results_dataset).await;
+
+    bench_spicecloud::run(&mut rt, &mut benchmark_results).await?;
+
+    let data_update: DataUpdate = benchmark_results.into();
+
+    let display_records = data_update.data.clone();
+
+    if let Some(upload_results_dataset) = upload_results_dataset {
+        tracing::info!("Writing benchmark results to dataset {upload_results_dataset}...");
+        setup::write_benchmark_results(data_update, &rt).await?;
+    }
+
+    display_benchmark_records(display_records).await?;
+
+    Ok(())
+}
+
+fn get_current_unix_ms() -> i64 {
+    let now = std::time::SystemTime::now();
+    now.duration_since(std::time::UNIX_EPOCH)
+        .map(|d| i64::try_from(d.as_millis()).unwrap_or(0))
+        .unwrap_or(0)
+}
+
+async fn run_query_and_record_result(
+    rt: &mut Runtime,
+    benchmark_results: &mut BenchmarkResultsBuilder,
+    query_name: &str,
+    query: &str,
+) -> Result<(), String> {
+    tracing::info!("Running query `{query_name}`...");
+    let start_time = get_current_unix_ms();
+
+    let mut min_iter_duration_ms = i64::MAX;
+    let mut max_iter_duration_ms = i64::MIN;
+
+    for _ in 0..benchmark_results.iterations() {
+        let start_iter_time = get_current_unix_ms();
+        let _ = rt
+            .df
+            .ctx
+            .sql(query)
```

Comment:
Should we use QueryBuilder w/ disabled cache to detect potential performance regression while we are adding more features to query?  he difference between two is that we have extra logic in QueryBuilder to collect table names, detect whether query is accelerated or not, etc. Or even have two runs - direct query and via QueryBuilder.

Reply:
We should have both cached and non-cached runs.

Reply:
I'll leave it as is for now.